modules/gtprof/trace_otel.go

```diff
@@ -0,0 +1,139 @@
+	Attributes []*OtelAttribute `json:"attributes,omitempty"`
+}
+
+type OtelScope struct {
+	Name       string           `json:"name"`
+	Version    string           `json:"version"`
+	Attributes []*OtelAttribute `json:"attributes,omitempty"`
+}
+
+type OtelSpan struct {
+	TraceID           string           `json:"traceId"`
+	SpanID            string           `json:"spanId"`
+	ParentSpanID      string           `json:"parentSpanId,omitempty"`
+	Name              string           `json:"name"`
+	StartTimeUnixNano string           `json:"startTimeUnixNano"`
+	EndTimeUnixNano   string           `json:"endTimeUnixNano"`
+	Kind              int              `json:"kind"`
+	Attributes        []*OtelAttribute `json:"attributes,omitempty"`
+}
+
+type OtelScopeSpan struct {
+	Scope *OtelScope  `json:"scope"`
+	Spans []*OtelSpan `json:"spans"`
+}
+
+type OtelResourceSpan struct {
+	Resource   *OtelResource    `json:"resource"`
+	ScopeSpans []*OtelScopeSpan `json:"scopeSpans"`
+}
+
+type OtelTrace struct {
+	ResourceSpans []*OtelResourceSpan `json:"resourceSpans"`
+}
+
+type OtelExporter struct{}
+
+func (e *OtelExporter) toOtelSpan(traceID string, t *traceBuiltinSpan, scopeSpan *OtelScopeSpan) {
+	t.ts.mu.RLock()
+	defer t.ts.mu.RUnlock()
+
+	span := &OtelSpan{
+		TraceID:           traceID,
+		SpanID:            t.ts.id,
+		Name:              t.ts.name,
+		StartTimeUnixNano: strconv.FormatInt(t.ts.startTime.UnixNano(), 10),
+		EndTimeUnixNano:   strconv.FormatInt(t.ts.endTime.UnixNano(), 10),
+		Kind:              2,
+	}
+
+	if t.ts.parent != nil {
+		span.ParentSpanID = t.ts.parent.id
+	}
+
+	scopeSpan.Spans = append(scopeSpan.Spans, span)
+
+	for _, a := range t.ts.attributes {
+		var otelVal any
+		if a.Value.IsString() {
+			otelVal = OtelAttributeStringValue{a.Value.AsString()}
+		} else {
+			otelVal = OtelAttributeIntValue{strconv.FormatInt(a.Value.AsInt64(), 10)}
+		}
+		span.Attributes = append(span.Attributes, &OtelAttribute{Key: a.Key, Value: otelVal})
+	}
+
+	for _, c := range t.ts.children {
+		child := c.internalSpans[t.internalSpanIdx].(*traceBuiltinSpan)
+		e.toOtelSpan(traceID, child, scopeSpan)
+	}
+}
+
+func (e *OtelExporter) recordTrace(t *traceBuiltinSpan) {
+	var spans []*OtelSpan
+
+	scopeSpan := &OtelScopeSpan{
+		Scope: &OtelScope{
+			Name:    "gitea-web",
+			Version: "1.0", // FIXME: get version from somewhere
+		},
+		Spans: spans,
+	}
+
+	traceID := GetTracer().randomHexForBytes(16)
+	e.toOtelSpan(traceID, t, scopeSpan)
+
+	resSpans := OtelResourceSpan{
+		Resource: &OtelResource{
+			Attributes: []*OtelAttribute{
+				{Key: "service.name", Value: OtelAttributeStringValue{"gitea"}},
+			},
+		},
+		ScopeSpans: []*OtelScopeSpan{scopeSpan},
+	}
+
+	otelTrace := &OtelTrace{ResourceSpans: []*OtelResourceSpan{&resSpans}}
+
+	// TODO: use a async queue
+	otelTraceJSON, err := json.Marshal(otelTrace)
+	if err == nil {
+		_, _ = http.Post("http://localhost:4318/v1/traces", "application/json", bytes.NewReader(otelTraceJSON))
```

Comment:
I know this is just a WIP so far, but would you be able to use the Gitea httpclient so it can use default settings such as a timeout as in case localhost isn't available we don't want so many pending connections open.

Reply:
Yes, it is WIP (there are some non-WIP PRs like `https://github.com/go-gitea/gitea/pull/33901`)

And I guess I won't use Gitea httpclient but use a fully customized one for this purpose.

Reply:
Awesome, thanks! Figured you had thought of it already, but wanted to err on the side of caution and mention it just in case:)